### PR TITLE
docs: add defined schema docs

### PR DESCRIPTION
**Pull request by Moumouls**
Add defined schema docs.

See #872

### CODE REVIEW

**Comment by Moumouls** [comment]
@sadortun  if you want to take a look, feel to suggest some commit changes from github UI 🙂

File: _includes/defined-schema/class-level-permissions.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+- `find`: Control search permissions
+- `get`: Control direct ID get permission
+- `count`: Control counting objects permission
+- `create`: Create permission
+- `update`: Update permission
+- `delete`: Delete permission
+- `protectedFields`: Control get permission at field level
+
+You can set each CLP field with options to add a first strong security layer. This security layer will be applied on the Parse Class and all Parse Objects into this class.
```
Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md]: "into this class"? Do you mean sub classes?

File: _includes/defined-schema/class-level-permissions.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+Available options for CLP keys:
+
+- `role:MyRole`: If you have already created a Parse Role, you can use your created Parse Role (ie: `MyRole`) in CLP keys.
+- `requiresAuthentication`: If true an authenticated user will have the permission.
+- `*`: Everybody has the permission
+- `{}`: if you set the CLP key with `{}` like `create: {}` only calls with your Parse Server Master Key will have the permission
+
+## CLP Protected Fields Key
+
+This CLP key is powerful and need some additional explanation.
```
Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md]: I think this line can be removed, it doesn't seem to provide any relevant info. Maybe instead a sentence about what this is (used for).

File: _includes/defined-schema/class-level-permissions.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+```js
+{
+    protectedFields: {
+      "*": ["authData", "emailVerified", "password", "username"],
+    },
+}
+```
+
+Listed keys under `*` will be protected from all users. Don't worry by default, `authData`, `emailVerified`, `password` are protected.
```
Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md]: Remove "don't worry"

File: _includes/defined-schema/class-level-permissions.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+But here for example we protect `username` from all users. So user A, even authenticated will not be able to get the `username` of a user B.
```
Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md]: Where is "here"? There is an example before and one after.

Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md]: The references "user A", "user B" seem to not be used anywhere else? If so, just simplify this to something like "a user cannot ... of other users".

File: _includes/defined-schema/class-level-permissions.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+But protected fields could be combined for example.
```
Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md]: Make this is clear reference, like: "Protected fields ... as in this example:"

File: _includes/defined-schema/class-level-permissions.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+{
+    protectedFields: {
+      "*": ["authData", "emailVerified", "password", "username", "phone", "score"],
+      "role:Admin": ["password", "authData", "emailVerified"],
+      "role:VerifiedUser": ["password", "authData", "emailVerified", "score"],
+    },
+}
+```
+
+In this case, a user member of the Parse Role `Admin` will be able to get the `phone` and `score` of another User. A user member of the Parse Role `VerifiedUser` can only get `phone`.
```
Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md]: In which case, the example above?

File: _includes/defined-schema/core-classes-fields.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
+# Core Classes/Fields
+
+Parse will never delete these fields on **ALL** classes if not provided in a class schema
```
Comment (mtrezza) [inline _includes/defined-schema/core-classes-fields.md:3]: "will never delete the following fields from any class, even if these fields are not defined in a class schema"

File: _includes/defined-schema/fields.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
+# Fields
+
+You can find here all field types available on Parse
```
Comment (mtrezza) [inline _includes/defined-schema/fields.md]: "These field types are available..."

File: _includes/defined-schema/fields.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+- `Number`: this type support `required` and `defaultValue`
```
Comment (mtrezza) [inline _includes/defined-schema/fields.md]: Maybe male this a table? Or at least remove the repetitive terms.

File: _includes/defined-schema/getting-started.md [hunk 10]
```diff
@@ -0,0 +1,122 @@
+# Getting Started
+
+## Introduction
+
+Parse Server was historically designed to be a schema less system. You didn't have to perform migration and define specific database schemas at start up time.
```
Comment (mtrezza) [inline _includes/defined-schema/getting-started.md]: "schemaless", "start-up"

Remove "historically", it's confusing (is it still today designed that way?). Instead:
"parse server is designed to be..."

File: _includes/defined-schema/getting-started.md [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+But schema less do not fulfill each use case of developers and could also do not play very well with some new Parse Server features like the GraphQL API.
```
Comment (mtrezza) [inline _includes/defined-schema/getting-started.md]: Remove the 2 prose paragraphs, instead simplify to the facts:

"For use cases in which a pre-defined schema is beneficial or required, you can define class fields, indexes, Class Level Permissions and more."

File: _includes/defined-schema/getting-started.md [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+
+To meet these needs, Parse Server now introduce Defined Schemas feature.
+Define easily you Parse Classes fields, indexes, Class level permissions and more.
+
+## Quick Start
+
+To leverage the power of Defined Schema we recommend to setup parse user like an express app.
```
Comment (mtrezza) [inline _includes/defined-schema/getting-started.md]: Rephrase this, is it required or why do we "recommend"?

File: _includes/defined-schema/indexes.md [hunk 13]
```diff
@@ -0,0 +1,26 @@
+# Indexes
+
+To optimize you Parse Server database performance you can define indexes and compound indexes.
```
Comment (mtrezza) [inline _includes/defined-schema/indexes.md]: "To optimize the Parse Server performance..."

File: _includes/defined-schema/indexes.md [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+To define an index on a `Pointer` field you need to use a
+special notation `_p_myFieldName`.
```
Comment (mtrezza) [inline _includes/defined-schema/indexes.md]: `_p_<FIELDNAME>`

File: _includes/defined-schema/indexes.md [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+For example if you define `city: { type: "Pointer", targetClass: "City" }` in your `fields` you can define an index on this pointer with `cityIndexExample: { _p_city: true }`
+
+Note: Currently Defined Schemas do not support indexes on special `_Join` classes used under the hood by the `Relation` type
```
Comment (mtrezza) [inline _includes/defined-schema/indexes.md]: Dot at end

File: _includes/defined-schema/options.md [hunk 16]
```diff
@@ -0,0 +1,39 @@
+# Options
+
+## definitions
+
+You classes definitions stored in an `Array`.
```
Comment (mtrezza) [inline _includes/defined-schema/options.md]: Could you rephrase this?

File: _includes/defined-schema/options.md [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+
+## strict
+
+You can set the `strict` option to `true` if you want parse-server to delete removed classes from your schemas from your database. Data stored in removed classes will be lost.
```
Comment (mtrezza) [inline _includes/defined-schema/options.md]: "Parse Server"

Could you rephrase "from your schemas from your database", may be confusing

File: _includes/defined-schema/options.md [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+
+`strict` is default to `false`. If you often change your schemas be aware that you can have some stale classes in your database. You will need to delete these classes manually.
```
Comment (mtrezza) [inline _includes/defined-schema/options.md]: "By default `strict` is `false`.

"You can have stale..." sounds like it's permissible. I think you want to say that the stale classes are unwanted and may remain?

File: _includes/defined-schema/options.md [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+
+## deleteExtraFields
+
+You can set the `deleteExtraFields` option to `true` if you parse-server to delete removed a class field from your database. Data stored in the removed field will be lost.
```
Comment (mtrezza) [inline _includes/defined-schema/options.md]: "if you want Parse Server to..."

"Delete remove" - a typo?

File: _includes/defined-schema/options.md [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+
+`deleteExtraFields` is default to `false`. Be aware that some stale fields could exists in your database. You will need to delete these fields manually.
```
Comment (mtrezza) [inline _includes/defined-schema/options.md]: Same here rephrasing

File: _includes/defined-schema/options.md [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+
+## lockSchemas
+
+You can set the `lockSchemas` option to `true` if you want to to prevent any `Parse.Schema` manipulation outside of the Defined Schema feature. If this options is `true` any create/update/delete request to `Parse.Schema` will be denied. You will not be able to manipulate `indexes`, `classLevelPermissions`, `fields`.
+
+This option help to keep one source of truth. And prevent developers or custom code to interfere with your schema definitions and data structure, even with the master key.
+
+## beforeMigration
+
+A function called before parse-server performs schema updates based on the `definitions` option
```
Comment (mtrezza) [inline _includes/defined-schema/options.md:35]: Always use "Parse Server" when you refer to the product, don't use "parse-server" unless you are referring to the repository.

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: _includes/defined-schema/class-level-permissions.md [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+- `create`: Create permission
+- `update`: Update permission
+- `delete`: Delete permission
+- `protectedFields`: Control get permission at field level
+
+You can set each CLP field with options to add a first strong security layer. This security layer will be applied on the Parse Class and all Parse Objects into this class.
+
+Note: If you update CLP you do not need to update Parse Objects. CLP is a security layer at Class Level not Object Level. For Object Level permission you can take a look to ALCs. You can use CLPs combined to ACLs to deeply secure your Parse Server.
+
+## CLP Key Options
```
Comment (sadortun) [inline _includes/defined-schema/class-level-permissions.md]: I would suggest replacing globally `key` with either `parameter` or `attribute` depending on the context.

Currently `parameter` seems to be more generalized as config keys

https://docs.parseplatform.org/parse-server/guide/#usage

File: _includes/defined-schema/getting-started.md [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+    lockSchemas: true,
+    // If true, Parse Server will delete non defined Classes from
+    // the database. (Core classes like Role, User are never deleted)
+    strict: true,
+    // If true, a field type change, the changed field is deleted
+    // from the database (all data in this field will be deleted)
+    // and then create the field with the new type
+    recreateModifiedFields: true,
+    // If true, Parse will delete non defined fields on a class. (Core fields are never deleted)
+    deleteExtraFields: true,
```
Comment (sadortun) [inline _includes/defined-schema/getting-started.md]: Theses 2 parameters should be `false` by default in the example to avoir a user copy-pasting the example config have it's data deleted unintentionally.

**Comment by Moumouls** [comment]
@sadortun  @mtrezza  requested changes performed. Feel free to use the Github commit suggestion feature if you some new changes suggestions, then i can apply easily your changes :)

File: _includes/defined-schema/options.md [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+
+You can set the `deleteExtraFields` option to `true` if you want Parse Server to delete a removed Defined Parse Class field from your database. Data stored in the removed field will be lost.
+
+`deleteExtraFields` is default to `false`. Be aware that you will have stale data fields in your database since Parse Server will not delete field data automatically. You will need to delete these fields manually.
+
+## recreateModifiedFields
+
+You can set the `recreateModifiedFields` option to `true` if you want Parse Server to clean field data before Parse Server update the field type when you change the type of a field (ie: from `String` to `Number`). Data stored on the modified field will be lost.
+
+`recreateModifiedFields` is default to `false`. Be aware that if you do not perform some data migration, you can result with data type inconsistency on modified field.
```
Comment (sadortun) [inline _includes/defined-schema/options.md]: Maybe adding some Bold for ?

**Be aware that if you do not perform some data migration**

**Review by sadortun — COMMENTED** [review]
Excellent! 

Only thing could be to add some Bold/red font on dangerous options

**Comment by Moumouls** [comment]
@sadortun  i added bold, @mtrezza  since Parse V5 was released, what do you think about merging this one ?

File: _includes/defined-schema/class-level-permissions.md [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+Note: If you update CLP you do not need to update Parse Objects. CLP is a security layer at Class Level not Object Level. For Object Level permission you can take a look to ALCs. You can use CLPs combined to ACLs to deeply secure your Parse Server.
+
+## CLP Parameter Options
+
+Available options for CLP parameters:
+
+- `role:MyRole`: If you have already created a Parse Role, you can use your created Parse Role (ie: `MyRole`) in CLP keys.
+- `requiresAuthentication`: If true an authenticated user will have the permission.
+- `*`: Everybody has the permission
+- `{}`: if you set the CLP key with `{}` like `create: {}` only calls with your Parse Server Master Key will have the permission
```
Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md:28]: "\`role:\<roleName\>\`: If you are making use of Parse Roles you can set the permission based on a role."
...
"If you set the CLP key to \`{}\`" ... "calls with the Parse Server master key will"

File: _includes/defined-schema/class-level-permissions.md [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+You can set each CLP parameter to add a first strong security layer. This security layer will be applied on the Parse Class and will cover all Parse Objects of the Parse Class.
+
+Note: If you update CLP you do not need to update Parse Objects. CLP is a security layer at Class Level not Object Level. For Object Level permission you can take a look to ALCs. You can use CLPs combined to ACLs to deeply secure your Parse Server.
+
+## CLP Parameter Options
+
+Available options for CLP parameters:
+
+- `role:MyRole`: If you have already created a Parse Role, you can use your created Parse Role (ie: `MyRole`) in CLP keys.
+- `requiresAuthentication`: If true an authenticated user will have the permission.
```
Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md:26]: "If set to \`true\`, only authenticated users"

Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md:26]: "If set to \`true\`, only authenticated users"

File: _includes/defined-schema/class-level-permissions.md [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+- `*`: Everybody has the permission
+- `{}`: if you set the CLP key with `{}` like `create: {}` only calls with your Parse Server Master Key will have the permission
+
+## CLP Protected Fields Parameter
+
+This CLP parameter allows you to restrict access to fields to specific Parse users.
+
+We will take the Parse User Class as example.
```
Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md:34]: "as an example:"

Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md:34]: "as an example."

File: _includes/defined-schema/class-level-permissions.md [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+// className: '_User'
+{
+    protectedFields: {
+      "*": ["authData", "emailVerified", "password", "username"],
+    },
+}
+```
+
+Listed keys under `*` will be protected from all users. By default, `authData`, `emailVerified`, `password` are protected.
+But in the above example we protect `username` from all users. So a Parse User, even authenticated will not be able to get the `username` of a another Parse User.
```
Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md:46]: Remove next line between the 2 sentences.

Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md:46]: Remove next line between the 2 sentences.

File: _includes/defined-schema/class-level-permissions.md [hunk 29]
```diff
@@ -0,0 +1,61 @@
...
+
+`protectedFields` could be also combined as in the following example.
```
Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md:48]: "following example:"

Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md:48]: "following example:"

File: _includes/defined-schema/class-level-permissions.md [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+{
+    protectedFields: {
+      "*": ["authData", "emailVerified", "password", "username", "phone", "score"],
+      "role:Admin": ["password", "authData", "emailVerified"],
+      "role:VerifiedUser": ["password", "authData", "emailVerified", "score"],
+    },
+}
+```
+
+In the example above, a Parse User member of the Parse Role `Admin` will be able to get the `phone` and `score` of another Parse User. A Parse User member of the Parse Role `VerifiedUser` can only get `phone`.
```
Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md:60]: " A Parse User who is member of the Parse Role"

Comment (mtrezza) [inline _includes/defined-schema/class-level-permissions.md:60]: " A Parse User who is member of the Parse Role"

File: _includes/defined-schema/core-classes-fields.md (hunk 7 above)
Comment (mtrezza) [reply _includes/defined-schema/core-classes-fields.md:3]: still open

File: _includes/defined-schema/fields.md [hunk 31]
```diff
@@ -0,0 +1,42 @@
+# Fields
+
+These field types are available on a Parse Schema.
+
+`required`: `boolean`, by default false. Force the field to be set on create and update.
```
Comment (mtrezza) [inline _includes/defined-schema/fields.md:5]: "by default \`false\`."

Comment (mtrezza) [inline _includes/defined-schema/fields.md:5]: ", forces the field to be set on create and update, is \`false\` by default."

File: _includes/defined-schema/getting-started.md [hunk 32]
```diff
@@ -0,0 +1,118 @@
...
+    lastname: { type: "String", required: true },
+    tags: { type: "Array" },
+    location: { type: "GeoPoint" },
+    city: { type: "Pointer", targetClass: "City" },
+    friends: { type: "Relation", targetClass: "_User" },
+    zone: { type: "Polygon" },
+  },
+  indexes: {
+    tagsIndex: { tags: 1 },
+    // Special _p_ word to create indexes on pointer fields
```
Comment (mtrezza) [inline _includes/defined-schema/getting-started.md:28]: "// The special prefix \_p\_ is used to create indexes on pointer fields"

Comment (mtrezza) [inline _includes/defined-schema/getting-started.md:28]: "// Special _p_ prefix"

File: _includes/defined-schema/getting-started.md [hunk 33]
```diff
@@ -0,0 +1,118 @@
...
+const Country = {
+  className: "Country",
+  fields: {
+    name: { type: "String", required: true },
+  },
+  classLevelPermissions: {
+    find: { requiresAuthentication: true },
+    count: { requiresAuthentication: true },
+    get: { requiresAuthentication: true },
+    // Empty object meas that only master key
```
Comment (mtrezza) [inline _includes/defined-schema/getting-started.md:75]: write comment on 1 line instead of 2

Comment (mtrezza) [inline _includes/defined-schema/getting-started.md:75]: write comment on 1 line instead of 2

File: _includes/defined-schema/getting-started.md [hunk 34]
```diff
@@ -0,0 +1,118 @@
...
+};
+
+ParseServer.start({
+  databaseURI: "mongodb://your.mongo.uri",
+  appId: "myAppId",
+  masterKey: "mySecretMasterKey",
+  serverURL: "http://localhost:1337/parse",
+  port: 1337,
+  publicServerURL: "http://localhost:1337/parse",
+  // Then just register schemas into Parse Server
```
Comment (mtrezza) [inline _includes/defined-schema/getting-started.md:90]: // Define schemas of Parse Server

Comment (mtrezza) [inline _includes/defined-schema/getting-started.md:90]: // Define schemas of Parse Server

File: _includes/defined-schema/getting-started.md [hunk 35]
```diff
@@ -0,0 +1,118 @@
...
+  schema: {
+    definitions: [User, City, Country],
+    // Parse Schema API will be disabled
```
Comment (mtrezza) [inline _includes/defined-schema/getting-started.md:93]: // Disable Parse Schema API; if you need to update the schemas later
// you need to redeploy Parse Server with the new schema definition

Comment (mtrezza) [inline _includes/defined-schema/getting-started.md:93]: // If set to \`true\`, the Parse Server API for schema changes is disabled and schema
// changes are only possible by redeploying Parse Server with a new schema definition

File: _includes/defined-schema/getting-started.md [hunk 36]
```diff
@@ -0,0 +1,118 @@
...
+    // If you need to update schemas Parse server
+    // need to be updated and deployed (CI/CD strategy)
+    lockSchemas: true,
+    // If true, Parse Server will delete non defined Classes from
```
Comment (mtrezza) [inline _includes/defined-schema/getting-started.md:97]: // If set to \`true\`, Parse Server will automatically delete non-defined classes from
// the database; internal classes like \`User\` or \`Role\` are never deleted.

Comment (mtrezza) [inline _includes/defined-schema/getting-started.md:97]: // Parse Server will automatically delete non-defined classes from the
// database; internal classes like \`User\` or \`Role\` are never deleted.

File: _includes/defined-schema/getting-started.md [hunk 37]
```diff
@@ -0,0 +1,118 @@
...
+    // the database. (Core classes like Role, User are never deleted)
+    strict: true,
+    // If true, a field type change, the changed field is deleted
+    // from the database (all data in this field will be deleted)
+    // and then create the field with the new type
+    recreateModifiedFields: false,
```
Comment (mtrezza) [inline _includes/defined-schema/getting-started.md:103]: // If set to \`true\`, a field type change will cause the field including its data to be
// deleted from the database, and then a new field to be created with the new type

Comment (mtrezza) [inline _includes/defined-schema/getting-started.md:103]: // If set to \`true\`, a field type change will cause the field including all of its data
// being deleted from the database and then recreate the field with the new type

File: _includes/defined-schema/getting-started.md [hunk 38]
```diff
@@ -0,0 +1,118 @@
...
+    // If true, Parse will delete non defined fields on a class. (Core fields are never deleted)
+    deleteExtraFields: false,
```
Comment (mtrezza) [inline _includes/defined-schema/getting-started.md:105]: // If set to \`true\`, Parse Server will automatically delete non-defined class fields;
// internal fields in classes like `User` or `Role` are never deleted.

Comment (mtrezza) [inline _includes/defined-schema/getting-started.md:105]: // If set to \`true\`, Parse Server will automatically delete non-defined class fields.
// internal fields in classes like `User` or `Role` are never deleted.

File: _includes/defined-schema/getting-started.md [hunk 39]
```diff
@@ -0,0 +1,118 @@
...
+  },
+  serverStartComplete: () => {
+    // Here your Parse Server is ready
+    // with schemas up to date
+
+    // Just a code example if you want to expose
+    // an endpoint when parse is fully initialized
+    parseServer.expressApp.get("/ready", (req: any, res: any) => {
```
Comment (mtrezza) [inline _includes/defined-schema/getting-started.md:113]: // Parse Server is ready with up-to-date schema

Comment (mtrezza) [inline _includes/defined-schema/getting-started.md:113]: Replace whole comment block with:

// Parse Server is ready with up-to-date schema

File: _includes/defined-schema/indexes.md [hunk 40]
```diff
@@ -0,0 +1,26 @@
+# Indexes
+
+To optimize the Parse Server performance you can define indexes and compound indexes.
+
+To define an index on a `Pointer` field you need to use a
```
Comment (mtrezza) [inline _includes/defined-schema/indexes.md:5]: Reformat and reword to:

To optimize the Parse Server performance you can define indexes and compound indexes. Parse Server does not support indexes on special `_Join` classes used under the hood by the `Relation` type.

To define an index on a `Pointer` field you need to use the special notation `_p_<FIELDNAME>`. For example, if you define `city: { type: "Pointer", targetClass: "City" }` in your `fields` you can define an index on this pointer with `cityIndexExample: { _p_city: true }`.

Comment (mtrezza) [inline _includes/defined-schema/indexes.md:5]: Reformat and reword to:

To optimize the Parse Server performance you can define indexes and compound indexes. Parse Server does not support indexes on special `_Join` classes used under the hood by the `Relation` type.

To define an index on a `Pointer` field you need to use the special notation `_p_<FIELDNAME>`. For example, if you define `city: { type: "Pointer", targetClass: "City" }` in your `fields` you can define an index on this pointer with `cityIndexExample: { _p_city: true }`.

File: _includes/defined-schema/options.md [hunk 41]
```diff
@@ -0,0 +1,39 @@
...
+
+## deleteExtraFields
+
+You can set the `deleteExtraFields` option to `true` if you want Parse Server to delete a removed Defined Parse Class field from your database. Data stored in the removed field will be lost.
+
+`deleteExtraFields` is default to `false`. Be aware that you will have stale data fields in your database since Parse Server will not delete field data automatically. You will need to delete these fields manually.
+
+## recreateModifiedFields
+
+You can set the `recreateModifiedFields` option to `true` if you want Parse Server to clean field data before Parse Server update the field type when you change the type of a field (ie: from `String` to `Number`). Data stored on the modified field will be lost.
```
Comment (mtrezza) [inline _includes/defined-schema/options.md:21]: "before updating the field type when you change the field type, for example from `String` to `Number`. Data stored under the modified"

Comment (mtrezza) [inline _includes/defined-schema/options.md:21]: "before updating the field type when you change the field type"

File: _includes/defined-schema/options.md [hunk 42]
```diff
@@ -0,0 +1,39 @@
...
+
+`recreateModifiedFields` is default to `false`. **Be aware that if you do not perform some data migration, you can result with data type inconsistency on modified field.**
```
Comment (mtrezza) [inline _includes/defined-schema/options.md:23]: \`recreateModifiedFields\` defaults to \`false\`.

The "Be aware" comment could be unclear for readers, could you rephrase that? What is "some data migration"? In which cases does the user have to be aware, if the field is true of false?

Comment (mtrezza) [inline _includes/defined-schema/options.md:23]: \`recreateModifiedFields\` defaults to \`false\`

File: _includes/defined-schema/options.md [hunk 43]
```diff
@@ -0,0 +1,39 @@
...
+
+On production a good practice could be to create a new field with your new type, and then create a Parse Cloud Job to migrate old field data to the new created field.
```
Comment (mtrezza) [inline _includes/defined-schema/options.md:25]: "Good practice would be to create the new field of the new type" ... "to the newly created field."

Comment (mtrezza) [inline _includes/defined-schema/options.md:25]: "Good practice would be to create a new field with a new type" ... "to the newly created field."

File: _includes/defined-schema/options.md [hunk 44]
```diff
@@ -0,0 +1,39 @@
...
+
+## lockSchemas
+
+You can set the `lockSchemas` option to `true` if you want to to prevent any `Parse.Schema` manipulation outside of the Defined Schema feature. If this options is `true` any create/update/delete request to `Parse.Schema` will be denied. You will not be able to manipulate `indexes`, `classLevelPermissions`, `fields`.
```
Comment (mtrezza) [inline _includes/defined-schema/options.md:29]: "if you want to prevent any schema manipulation and to lock the schema as defined in the Parse Server configuration." ... "any create, update and delete request will be denied by the Parse Server API, even with the master key."

You can delete the following sentence "This option help to keep " as it does not provide any additional information.

Comment (mtrezza) [inline _includes/defined-schema/options.md:29]: "if you want to prevent any schema manipulation and to lock the schema as defined in the Parse Server configuration."

File: _includes/defined-schema/options.md (hunk 21 above)
Comment (mtrezza) [reply _includes/defined-schema/options.md:35]: still open

Comment (mtrezza) [reply _includes/defined-schema/options.md:35]: still open

**Review by mtrezza — CHANGES_REQUESTED** [review]
General feedback: the texts are quite verbose. Try to condense essential information into few words. Try to use as few lines as possible, instead of many lines with only a few words per line, that makes it easier to read and uses less space. For example:

```
// Here your Parse Server is ready
// with schemas up to date

// Just a code example if you want to expose
// an endpoint when parse is fully initialized
```
can be condensed into
```
// Parse Server is ready with up-to-date schema
```

**Comment by stephannielsen** [comment]
Maybe not the best place to ask but as there is no official documentation available yet:

* Can we define on schema levels Array of Pointers of a specific class? This PR does not mention this and from some tests it seems that if you specify `Array` as type in the schema, anything is accepted as array content.
* It would be nice if the documentation could include some examples on how to use `beforeMigration` and `afterMigration` hooks. I am not sure how I could leverage them to migrate data.

**Comment by mtrezza** [comment]
Thanks for the feedback to the docs, maybe @Moumouls can add this to the PR?
To your first point, the best place to ask is our [community forum](https://community.parseplatform.org).

**Comment by Moumouls** [comment]
thanks, @stephannielsen  for your feedback I'll quickly answer to your question. Parse is not currently able to type the content of `Array`. Anything could be stored in the `Array`. You can use the beforeSave Parse Cloud hook to prevent users saving anything other than the desired pointers.

If you use the GraphQL API. You can use the "inline fragment" graphql feature to automatically get data from the pointer array. 

https://community.parseplatform.org/t/type-arrayresult-must-have-a-selection-of-subfields-error/847/2?u=moumouls

for example `... on MyPointerClass { id aField}`

**Comment by stephannielsen** [comment]
Thanks @Moumouls and @mtrezza  for your quick replies.

I already expected that Parse can't type array fields and that triggers would allow me to validate incoming data. Thanks for the quick confirmation.

**Comment by RahulLanjewar93** [comment]
Hey, just had a quick question, since the documentation isn't updated yet. Can we have unique indexes according to the recent merge?

Thanks.

**Comment by mtrezza** [comment]
If anyone wants to pick up and finish this PR (see the review comments, just grammar and style corrections), please feel free to do so, so we can merge this chapter about defined schema into the docs. You will probably have to open a new PR, then we can close this one.

I have added a description to this PR and created issue #872 to track this.

**Comment by Moumouls** [comment]
@mtrezza  any chance to update your comments using the commit suggestion feature ? then I'll be happy to apply all your suggestions 🚀

**Comment by Moumouls** [comment]
closing in favor of https://github.com/parse-community/docs/pull/912
